compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -234,12 +228,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected_ty_expr: Option<&'tcx hir::Expr<'tcx>>,
         allow_two_phase: AllowTwoPhase,
     ) -> Ty<'tcx> {
-        let (ty, err) =
-            self.demand_coerce_diag(expr, checked_ty, expected, expected_ty_expr, allow_two_phase);
-        if let Some(err) = err {
-            err.emit();
+        match self.demand_coerce_diag(expr, checked_ty, expected, expected_ty_expr, allow_two_phase)
+        {
+            Ok(ty) => ty,
+            Err(err) => {
+                err.emit();
+                expected
```

Comment:
interesting that we return `expected` instead of `Ty::new_error` here :thinking: if someone wants to experiment with this, try changing it and run our UI test suite.

Reply:
I did have that, the result is not great. I should document it here